pkg/common/namespace_selection.go

```diff
@@ -103,3 +116,199 @@ func GetAllNamespaces(client kubernetes.Interface, labelSelector metav1.LabelSel
 
 	return namespacesNames, nil
 }
+
+type SelectorReconciler interface {
+	Get(string, policyv1.Target) ([]string, error)
+	HasUpdate(string) bool
+	Stop(string)
+}
+
+type NamespaceSelectorReconciler struct {
+	Client     client.Client
+	selections map[string]NamespaceSelection
+	lock       sync.RWMutex
+}
+
+type NamespaceSelection struct {
+	target     policyv1.Target
+	namespaces []string
+	hasUpdate  bool
+	err        error
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *NamespaceSelectorReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	r.selections = make(map[string]NamespaceSelection)
+
+	neverEnqueue := predicate.NewPredicateFuncs(func(o client.Object) bool { return false })
+
+	// Instead of reconciling for each Namespace, just reconcile once
+	// - that reconcile will do a list on all the Namespaces anyway.
+	mapToSingleton := func(_ client.Object) []reconcile.Request {
+		return []reconcile.Request{{NamespacedName: types.NamespacedName{
+			Name: "NamespaceSelector",
+		}}}
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("NamespaceSelector").
+		For( // This is a workaround because a `For` is required, but doesn't allow the enqueueing to be customized
+			&corev1.Namespace{},
+			builder.WithPredicates(neverEnqueue)).
+		Watches(
+			&source.Kind{Type: &corev1.Namespace{}},
+			handler.EnqueueRequestsFromMapFunc(mapToSingleton),
+			builder.WithPredicates(predicate.LabelChangedPredicate{})).
+		Complete(r)
+}
+
+func (r *NamespaceSelectorReconciler) Reconcile(ctx context.Context, _ ctrl.Request) (ctrl.Result, error) {
+	log := logf.Log.WithValues("Reconciler", "NamespaceSelector")
+	oldSelections := make(map[string]NamespaceSelection)
+
+	r.lock.RLock()
+
+	for name, selection := range r.selections {
+		oldSelections[name] = selection
+	}
+
+	r.lock.RUnlock()
+
+	namespaces := corev1.NamespaceList{}
+
+	// This List will be from the cache
+	if err := r.Client.List(ctx, &namespaces); err != nil {
```

Comment:
If `r.selections` is empty, could you please return before calling `List`? That way if there are no policies with namespaceSelector then you can avoid the memory allocation.